src/interpolation.jl

```diff
@@ -85,35 +85,69 @@ function interpolate_kpoint(data_in::AbstractVecOrMat, kpoint_in::Kpoint, kpoint
     data_out
 end
 
+"""
+Compute indices of the vectors in G_vectors(basis_in[ik]) that are also present
+in G_vectors(basis_out[ik]).
+"""
+function transfer_blochwave_mapping(ik, basis_in::PlaneWaveBasis{T},
+                                    basis_out::PlaneWaveBasis{T}) where T
+    # Get indices of the G vectors of the old basis inside the new basis.
+    idcsk_out = index_G_vectors.(Ref(basis_out), G_vectors(basis_in.kpoints[ik]))
+    # In the case where G_vectors(basis_in.kpoints[ik]) are biggers than vectors
+    # in the fft_size box of basis_out, we need to filter out the "nothings" to
+    # make sure that the indices linearization works. It is not an issue to
+    # filter these vectors as this can only happen if Ecut_in > Ecut_out.
+    filter!(!isnothing, idcsk_out)
+    # Linearize the indices
+    idcsk_out = getindex.(Ref(LinearIndices(basis_out.fft_size)), idcsk_out)
+
+    # Map to the indices of the corresponding G-vectors in G_vectors(kpt_out)
+    # this array might contains some nothings if basis_out has less G_vectors
+    # than basis_in at this kpoint
+    indexin(idcsk_out, basis_out.kpoints[ik].mapping)
+end
 
 """
-Interpolate Bloch wave between two basis sets. Limited feature set. Currently only
-interpolation to a bigger grid (larger Ecut) on the same lattice supported.
+Transfer Bloch wave between two basis sets. Limited feature set.
+For the moment, only PlaneWaveBasis with same lattice and kgrid are supported.
 """
-function interpolate_blochwave(ψ_in, basis_in, basis_out)
+function transfer_blochwave(ψ_in, basis_in::PlaneWaveBasis{T},
```

Comment:
I called this function `interpolate_` to resemble the similar structure to the density and kpoint functions above. Any reason why you decided to rename?

Reply:
because when Ecut_out < Ecut_in it's not really interpolation anymore :thinking:

Reply:
Ok that's true, but the same can be the case for `interpolate_kpoint` and we still call it `interpolate`. I personally don't mind which of the names we use, but it just feels strange to have one function called `interpolate_` and one `transfer_`. I guess transfer is more general and perhaps that should be used for all functions (and thus the file renamed to `basis_transfer.jl`). Thoughts?

Actually now that I think about that `transfer_blochwave` and `interpolate_kpoint` are actually quite related up to the point where one should probably merge these two ideas, no? In principle one could have a one-k-point-only version of `transfer_blochwave`, which could take two kpoint indices and two bases. We only use `interpolate_kpoint` once, where we can easily switch the interface from being `Kpoint`-based to being `PlaneWaveBasis`-based. OTOH perhaps a  `Kpoint`-based version of `transfer_blochwave` could also be possible and might be more general.

My point here is: I feel we are reinventing an existing concept, where we should generalise what we already have that it applies to both cases.

Reply:
I am not really sure that I see what you mean. You suggest to include `interpolate_kpoint` in `transfer_blochwave` rather than `transfer_blochwave_mapping` when it comes to transfer data from one k-point from the input basis to the output basis ?

Reply:
Personnaly I have no opinion about what is best :sweat_smile: as long as in the end it does what I need :stuck_out_tongue:

Reply:
Not quite ... sorry about the confusion. No what I mean is that `interpolate_kpoint` and `transfer_blochwave` are special cases of a more general function, which transfers data from one tuple (basis, kpoint) to another tuple (basis2, kpoint2) and I think with the infrastructure built in `interpolate_kpoint` and `transfer_blochwave` it should not be too hard and worth it to directly go for this general implementation.

Reply:
Yes sure, it's just I think it might we worth to go the extra mile now that we have these two special cases and then we don't need to worry about this any more in the future. At the very least we should unify both concepts in a function of the same name and related interface (if you don't have the time to generalise fully right now).

Reply:
Okay, it is clearer now, thanks ! 
Indeed it is worth it to do it now so that we don't have to worry about it in the future. I'll try to do this after dealing with Antoine reviews of newton PR ;)

Reply:
sure.I'll do another pass on that as well after you guys are done. Thanks for your work in both these PRs. Well appreciated!